README.md
<!-- Allow this file to not have a first line heading -->
<!-- markdownlint-disable-file MD041 -->

<!-- inline html -->
<!-- markdownlint-disable-file MD033 -->

<div align="center">

# `🔦 pdm-plugin-torch `

A utility tool for selecting torch backend and version.

[![Embark](https://img.shields.io/badge/embark-open%20source-blueviolet.svg)](https://embark.dev)
[![Embark](https://img.shields.io/badge/discord-ark-%237289da.svg?logo=discord)](https://discord.gg/dAuKfZS)
[![Build status](https://badge.buildkite.com/968ac3c0bb075fb878f9f973ed91406c8b257b0f050c197542.svg?theme=github&branch=main)](https://buildkite.com/embark-studios/pdm-plugin-torch)
[![Docs status](https://img.shields.io/badge/Docs-latest-brightgreen)](https://embarkstudios.github.io/pdm-plugin-torch/)
[![pdm-managed](https://img.shields.io/badge/PDM-v2.8.0-blueviolet)](https://pdm.fming.dev)

</div>


## What it does

Due to torch being published in many different variants with local versions to signify the underlying API, it is hard to integrate into a lockfile-based workflow. This is due to the local versions - `+cu111`, `+cpu`, and so on - being considered the "same" package, so you can't resolve two at the same time.

This tool generate multiple lockfiles *only for torch* and allows you to use a pdm subcommand (`pdm torch`) to select the one you want.

### Configuration

These are the supported options:

```toml
[tool.pdm.plugin.torch]
dependencies = [
   "torch==1.10.2"
]
lockfile = "torch.lock"
enable-cpu = true

enable-rocm = true
rocm-versions = ["4.2"]

enable-cuda = true
cuda-versions = ["cu111", "cu113"]
```

## Installation

PDM supports specifying plugin-dependencies in your pyproject.toml, which is the suggested installation method. Note that in `pdm-plugin-torch` versions before 23.4.0, our configuration was in `tool.pdm.plugins.torch`. If upgrading, you'll need to also change that to `tool.pdm.plugin.torch`.

``` toml
[tool.pdm]
plugins = ["pdm-plugin-torch==$VERSION"]
```

## Contribution

[![Contributor Covenant](https://img.shields.io/badge/contributor%20covenant-v1.4-ff69b4.svg)](../main/CODE_OF_CONDUCT.md)

We welcome community contributions to this project.

Please read our [Contributor Guide](CONTRIBUTING.md) for more information on how to get started.
Please also read our [Contributor Terms](CONTRIBUTING.md#contributor-terms) before you make any contributions.

Any contribution intentionally submitted for inclusion in an Embark Studios project, shall comply with the Rust standard licensing model (MIT OR Apache 2.0) and therefore be dual licensed as described below, without any additional terms or conditions:

### License

This contribution is dual licensed under EITHER OF

* Apache License, Version 2.0, ([LICENSE-APACHE](LICENSE-APACHE) or <http://www.apache.org/licenses/LICENSE-2.0>)
* MIT license ([LICENSE-MIT](LICENSE-MIT) or <http://opensource.org/licenses/MIT>)

at your option.

For clarity, "your" refers to Embark or any other licensee/user of the contribution.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [23.4.0] - 2023-11-14

- Configuration is now expected at `tool.pdm.plugin.torch`. Note the missing `s`. This is to avoid collision with the upstream configuration key.

## [23.3.0] - 2023-11-14

- Add support for PDM 2.9 and 2.10
- On all versions where applicable, enforces `static-url`.
- Removes support for PDM 2.8 as it has a show-stopper bug
- Fixed a bug with `CPU` target that would sometimes include non-CPU-specific wheels.

## [23.2.0] - 2023-09-14

- Drops support for PDM 2.3 and 2.4, adds support for PDM 2.6-2.8

## [23.1.1] - 2023-06-15

- Fix a bug where 2.5.0 would crash with an assertion.

## [23.1.0] - 2023-04-19

- Adds support for PDM 2.5.0

## [23.0.0] - 2023-03-01

This is the initial release

[Unreleased]: https://github.com/EmbarkStudios/pdm-plugin-torch/compare/23.3.0...HEAD
[23.3.0]: https://github.com/EmbarkStudios/pdm-plugin-torch/compare/23.2.0...23.3.0
[23.2.0]: https://github.com/EmbarkStudios/pdm-plugin-torch/compare/23.1.0...23.2.0
[23.1.1]: https://github.com/EmbarkStudios/pdm-plugin-torch/compare/23.1.0...23.1.1
[23.1.0]: https://github.com/EmbarkStudios/pdm-plugin-torch/compare/23.0.0...23.1.0
[23.0.0]: https://github.com/EmbarkStudios/pdm-plugin-torch/releases/tag/23.0.0


pdm-plugin-torch/pdm_plugin_torch/main.py
from __future__ import annotations

import sys

from typing import Iterable

import tomlkit

from pdm import __version__, termui
from pdm._types import RepositoryConfig
from pdm.cli.commands.base import BaseCommand
from pdm.cli.utils import fetch_hashes, format_lockfile, format_resolution_impossible
from pdm.core import Core
from pdm.models.candidates import Candidate
from pdm.models.repositories import BaseRepository, LockedRepository
from pdm.models.requirements import Requirement, parse_requirement
from pdm.models.specifiers import PySpecSet, get_specifier
from pdm.project import Project
from pdm.resolver import resolve
from pdm.resolver.providers import BaseProvider
from pdm.termui import Verbosity
from pdm.utils import atomic_open_for_write, expand_env_vars_in_auth
from resolvelib.reporters import BaseReporter
from resolvelib.resolvers import ResolutionImpossible, ResolutionTooDeep, Resolver

from pdm_plugin_torch.config import Configuration


is_pdm210 = PySpecSet(">=2.10").contains(__version__.__version__)
is_pdm29 = PySpecSet(">=2.9").contains(__version__.__version__)
is_pdm28 = PySpecSet(">=2.8").contains(__version__.__version__)


def sources(project: Project, sources: list) -> list[RepositoryConfig]:
    result: dict[str, RepositoryConfig] = {}
    for source in project.pyproject.settings.get("source", []):
        result[source["name"]] = RepositoryConfig(**source, config_prefix="pypi")

    for source in sources:
        result[source["name"]] = RepositoryConfig(**source, config_prefix="torch")

    def merge_sources(other_sources: Iterable[tuple[str, RepositoryConfig]]) -> None:
        for name, source in other_sources:
            source.name = name
            if name in result:
                result[name].passive_update(source)
            else:
                result[name] = source

    if not project.config.get("pypi.ignore_stored_index", False):
        if "pypi" not in result:  # put pypi source at the beginning
            result = {"pypi": project.default_source, **result}
        else:
            result["pypi"].passive_update(project.default_source)
        merge_sources(project.project_config.iter_sources())
        merge_sources(project.global_config.iter_sources())

    for source in result.values():
        assert source.url, "Source URL must not be empty"
        source.url = expand_env_vars_in_auth(source.url)

    return list(result.values())


def get_provider(
    project: Project,
    raw_sources: list,
    strategy: str = "all",
    for_install: bool = False,
    lockfile: dict = None,
    tracked_names: Iterable[str] | None = None,
    allow_prereleases: bool = False,
) -> BaseProvider:
    """Build a provider class for resolver.
    :param strategy: the resolve strategy
    :param tracked_names: the names of packages that needs to update
    :param for_install: if the provider is for install
    :returns: The provider object
    """
    from pdm.models.requirements import strip_extras
    from pdm.resolver.providers import (
        BaseProvider,
        EagerUpdateProvider,
        ReusePinProvider,
    )
    from pdm.utils import normalize_name

    repository = get_repository(
        project, raw_sources, for_install=for_install, lockfile=lockfile
    )

    overrides = {
        normalize_name(k): v for k, v in project.pyproject.resolution_overrides.items()
    }

    locked_repository: LockedRepository | None = None
    if strategy != "all" or for_install:
        try:
            locked_repository = LockedRepository(lockfile, sources, project.environment)
        except Exception:
            if for_install:
                raise
            project.core.ui.echo(
                "Unable to reuse the lock file as it is not compatible with PDM",
                style="warning",
                err=True,
            )

    if locked_repository is None:
        return BaseProvider(repository, allow_prereleases, overrides)

    if for_install:
        return BaseProvider(locked_repository, allow_prereleases, overrides)

    provider_class = ReusePinProvider if strategy == "reuse" else EagerUpdateProvider
    tracked_names = [strip_extras(name)[0] for name in tracked_names or ()]

    return provider_class(
        locked_repository.all_candidates,
        tracked_names,
        repository,
        allow_prereleases,
        overrides,
    )


def get_repository(
    project: Project,
    raw_sources: list,
    cls: type[BaseRepository] | None = None,
    for_install: bool = False,
    lockfile: dict = None,
) -> BaseRepository:
    """Get the repository object"""
    if cls is None:
        cls = project.core.repository_class

    fixed_sources = sources(project, raw_sources)
    return cls(
        fixed_sources,
        project.environment,
    )


def do_lock(
    project: Project,
    raw_sources: list,
    strategy: str = "all",
    requirements: list[Requirement] | None = None,
) -> dict[str, Candidate]:
    """Performs the locking process and update lockfile."""

    provider = get_provider(project, raw_sources, strategy)
    resolve_max_rounds = int(project.config["strategy.resolve_max_rounds"])
    ui = project.core.ui
    with ui.logging("lock"):
        # The context managers are nested to ensure the spinner is stopped before
        # any message is thrown to the output.
        try:
            with ui.open_spinner(title="Resolving dependencies") as spin:
                reporter = project.get_reporter(requirements, None, spin)
                resolver: Resolver = project.core.resolver_class(provider, reporter)
                mapping, dependencies = resolve(
                    resolver,
                    requirements,
                    project.environment.python_requires,
                    resolve_max_rounds,
                )

                spin.update("Fetching hashes for resolved packages...")
                fetch_hashes(provider.repository, mapping)

        except ResolutionTooDeep:
            ui.echo(f"{termui.Emoji.LOCK} Lock failed", err=True)
            ui.echo(
                "The dependency resolution exceeds the maximum loop depth of "
                f"{resolve_max_rounds}, there may be some circular dependencies "
                "in your project. Try to solve them or increase the "
                f"[green]`strategy.resolve_max_rounds`[/] config.",
                err=True,
            )
            raise
        except ResolutionImpossible as err:
            ui.echo(f"{termui.Emoji.LOCK} Lock failed", err=True)
            ui.echo(format_resolution_impossible(err), err=True)
            raise ResolutionImpossible("Unable to find a resolution") from None
        else:
            if is_pdm210:
                from pdm.project.lockfile import FLAG_STATIC_URLS

                data = format_lockfile(
                    project,
                    mapping,
                    dependencies,
                    groups=[],
                    strategy={FLAG_STATIC_URLS},
                )

            elif is_pdm29:
                data = format_lockfile(project, mapping, dependencies, static_urls=True)

            elif is_pdm28:
                data = format_lockfile(project, mapping, dependencies, static_urls=True)

            else:
                data = format_lockfile(project, mapping, dependencies)

            ui.echo(f"{termui.Emoji.LOCK} Lock successful")
            return data


def write_lockfile(
    project: Project, lock_name: str, toml_data: dict, show_message: bool = True
) -> None:
    toml_data["metadata"] = project.get_lock_metadata()
    lockfile_file = project.root / lock_name

    with atomic_open_for_write(lockfile_file) as fp:
        tomlkit.dump(toml_data, fp)  # type: ignore
    if show_message:
        project.core.ui.echo(f"Torch locks are written to [success]{lockfile_file}[/].")


def resolve_candidates_from_lockfile(
    project: Project,
    requirements: Iterable[Requirement],
    raw_sources,
    lockfile: dict,
) -> dict[str, Candidate]:
    ui = project.core.ui
    resolve_max_rounds = int(project.config["strategy.resolve_max_rounds"])
    reqs = [
        req
        for req in requirements
        if not req.marker or req.marker.evaluate(project.environment.marker_environment)
    ]
    with ui.logging("install-resolve"):
        with ui.open_spinner("Resolving packages from lockfile...") as spinner:
            reporter = BaseReporter()
            provider = get_provider(
                project, raw_sources, for_install=True, lockfile=lockfile
            )
            resolver: Resolver = project.core.resolver_class(provider, reporter)
            mapping, *_ = resolve(
                resolver,
                reqs,
                project.environment.python_requires,
                resolve_max_rounds,
            )
            spinner.update("Fetching hashes for resolved packages...")
            fetch_hashes(provider.repository, mapping)

    return mapping


def do_sync(
    project: Project,
    *,
    raw_sources: list,
    requirements: list[Requirement] | None = None,
    lockfile: dict,
) -> None:
    """Synchronize project"""

    candidates = resolve_candidates_from_lockfile(
        project, requirements, raw_sources, lockfile
    )

    handler = project.core.synchronizer_class(
        candidates,
        project.environment,
        clean=False,
        dry_run=False,
        no_editable=True,
        install_self=False,
        reinstall=False,
        only_keep=False,
        fail_fast=True,
    )

    with project.core.ui.logging("install"):
        handler.synchronize()


def read_lockfile(project: Project, lock_name: str) -> None:
    lockfile_file = project.root / lock_name

    data = tomlkit.parse(lockfile_file.read_text("utf-8"))
    return data


def is_lockfile_compatible(project: Project, lock_name: str) -> bool:
    lockfile_file = project.root / lock_name
    if not lockfile_file.exists():
        return True

    lockfile = read_lockfile(project, lock_name)
    lockfile_version = str(lockfile.get("metadata", {}).get("lock_version", ""))
    if not lockfile_version:
        return False

    if "." not in lockfile_version:
        lockfile_version += ".0"

    accepted = get_specifier(f"~={lockfile_version}")
    return accepted.contains(project.lockfile.spec_version)


def is_lockfile_hash_match(project: Project, lock_name: str) -> bool:
    lockfile_file = project.root / lock_name
    if not lockfile_file.exists():
        return False

    lockfile = read_lockfile(project, lock_name)
    hash_in_lockfile = str(lockfile.get("metadata", {}).get("content_hash", ""))
    if not hash_in_lockfile:
        return False

    algo, hash_value = hash_in_lockfile.split(":")
    content_hash = project.pyproject.content_hash(algo)

    return content_hash == hash_value


def check_lockfile(project: Project, lock_name: str) -> str | None:
    """Check if the lock file exists and is up to date. Return the update strategy."""
    lockfile_file = project.root / lock_name
    if not lockfile_file.exists():
        project.core.ui.echo("Lock file does not exist", style="warning", err=True)
        return False
    elif not is_lockfile_compatible(project, lock_name):
        project.core.ui.echo(
            "Lock file version is not compatible with PDM, installation may fail",
            style="yellow",
            err=True,
        )
        return False
    elif not is_lockfile_hash_match(project, lockfile_file):
        project.core.ui.echo(
            "Lock file hash doesn't match pyproject.toml, packages may be outdated",
            style="yellow",
            err=True,
        )
        return False
    return True


def get_settings(project: Project):
    return project.pyproject.settings["plugin"]["torch"]


class InstallCommand(BaseCommand):
    name = "install"
    description = "Install torch packages from lockfile"

    def add_arguments(self, parser):
        parser.add_argument("api", help="the api to use, e.g. cuda version or rocm")

    def handle(self, project: Project, options: dict):
        plugin_config = Configuration.from_toml(get_settings(project))

        resolves = plugin_config.variants
        if options.api not in resolves:
            raise ValueError(
                f"unknown API {options.api}, expected one of {[v for v in resolves]}"
            )

        lockfile = read_lockfile(project, plugin_config.lockfile)

        spec_for_version = lockfile[options.api]

        (source, local_version) = resolves[options.api]

        if is_pdm210:
            from pdm.project.lockfile import FLAG_STATIC_URLS

            class OverrideLockfile:
                def __init__(self, lockfile):
                    self._lockfile = lockfile

                @property
                def strategy(self):
                    strategies = self._lockfile.strategy
                    strategies.add(FLAG_STATIC_URLS)

                    return strategies

                def __getattr__(self, name):
                    return getattr(self._lockfile, name)

            original_lockfile = project.lockfile

            project._lockfile = OverrideLockfile(original_lockfile)

        reqs = [
            parse_requirement(f"{req}{local_version}", False)
            for req in plugin_config.dependencies
        ]

        do_sync(
            project,
            raw_sources=[
                {
                    "name": "torch",
                    "url": source,
                    "type": "index",
                    "verify_ssl": True,
                }
            ],
            requirements=reqs,
            lockfile=spec_for_version,
        )

        if is_pdm210:
            project._lockfile = original_lockfile


class LockCommand(BaseCommand):
    name = "lock"
    description = "Lock Torch and its dependencies to a specific version"

    def add_arguments(self, parser):
        parser.add_argument(
            "--check",
            help="validate that the lockfile is up to date",
            action="store_true",
        )

    def handle(self, project: Project, options: dict):
        plugin_config = Configuration.from_toml(get_settings(project))

        if options.check:
            is_updated = check_lockfile(project, plugin_config.lockfile)
            if not is_updated:
                project.core.ui.echo(
                    "Lockfile is [error]out of date[/].",
                    err=True,
                    verbosity=Verbosity.DETAIL,
                )
                sys.exit(1)
            else:
                project.core.ui.echo(
                    "Lockfile is [success]up to date[/].",
                    err=True,
                    verbosity=Verbosity.DETAIL,
                )
                sys.exit(0)

        results = {}
        for api, (url, local_version) in plugin_config.variants.items():
            reqs = [
                parse_requirement(f"{req}{local_version}", False)
                for req in plugin_config.dependencies
            ]

            results[api] = do_lock(
                project,
                [
                    {
                        "name": "torch",
                        "url": url,
                        "type": "index",
                        "verify_ssl": True,
                    }
                ],
                requirements=reqs,
            )

        write_lockfile(project, plugin_config.lockfile, results)


class TorchCommand(BaseCommand):
    """Generate a lockfile for torch specifically."""

    name = "torch"
    description = "Manage torch dependencies"

    def add_arguments(self, parser):
        subparsers = parser.add_subparsers(
            title="Sub commands", help="sub-command help", dest="command"
        )

        LockCommand.register_to(subparsers)
        InstallCommand.register_to(subparsers)

        self.parser = parser

    def handle(self, project: Project, options) -> None:
        self.parser.print_help()


def torch_plugin(core: Core):
    if is_pdm28 and not is_pdm29:
        raise RuntimeError(
            "pdm 2.8.* is not supported due to not https://github.com/pdm-project/pdm/issues/2151"
        )
    core.register_command(TorchCommand, "torch")
